fix(apprunner-alpha): env vars and secrets can't solely be added via .add*() methods

rogerchi opened the pull request:
This fixes the logic for rendering environment variables and environment secrets for the `apprunner-alpha` module.  Previously, `.addEnvironmentVariable()` and `.addSecret()` were being ignored if there were not already "seed" values in the input props.

Closes #24345.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation reviewed (dismissed):
The pull request linter has failed. See the aws-cdk-automation comment below for failure reasons. If you believe this pull request should receive an exemption, please comment and provide a justification.


A comment requesting an exemption should contain the text `Exemption Request`. Additionally, if clarification is needed add `Clarification Request` to a comment.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Fixes must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 74f6e462c87062607c3950f7cd7ee064d1150b00
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=029404dc-ee05-4aec-9921-8954423993b1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra commented:
Please make sure that your PR title confirms to the conventional commit standard (fix, feat, chore) and that it is written in a style that will reflect correctly in the change log (See [Contributing Guide, Pull Requests](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-4-pull-request))

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 0945f0f77d5850b0cce494f57639feedd5fad0ed
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0dd8d4de-9184-4029-a828-1d67871f9ae9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

rogerchi commented:
> Please make sure that your PR title confirms to the conventional commit standard (fix, feat, chore) and that it is written in a style that will reflect correctly in the change log (See [Contributing Guide, Pull Requests](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-4-pull-request))

I've updated the code, tests, and added a new integration test and snapshot.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: ed5eb0a4103c9e33ead0e6c313da32cb0f0bb530
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9194d08f-2965-4e94-9e56-17efe433fbd5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra commented:
> > Please make sure that your PR title confirms to the conventional commit standard (fix, feat, chore) and that it is written in a style that will reflect correctly in the change log (See [Contributing Guide, Pull Requests](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-4-pull-request))
> 
> I've updated the code, tests, and added a new integration test and snapshot.

I'll take a look but please also address the message above.

rogerchi commented:
> I'll take a look but please also address the message above.

Let me know if the new title works.

comcalvi commented on packages/@aws-cdk/aws-apprunner/test/service.test.ts:
```diff
@@ -391,6 +391,76 @@ test('custom environment secrets and start commands are allowed for imageConfigu
   });
 });
 
+
```
```suggestion
```

comcalvi commented on packages/@aws-cdk/aws-apprunner/test/service.test.ts:
```diff
@@ -391,6 +391,76 @@ test('custom environment secrets and start commands are allowed for imageConfigu
+  service.addSecret('LATER_SECRET', apprunner.Secret.fromSecretsManager(secret, 'field'));
+
+  // THEN
+  // we should have the service
```
```suggestion
```

comcalvi commented on packages/@aws-cdk/aws-apprunner/test/service.test.ts:
```diff
@@ -391,6 +391,76 @@ test('custom environment secrets and start commands are allowed for imageConfigu
+  // GIVEN
+  const app = new cdk.App();
+  const stack = new cdk.Stack(app, 'demo-stack');
+  // WHEN
+  const secret = new secretsmanager.Secret(stack, 'Secret');
+  const service = new apprunner.Service(stack, 'DemoService', {
+    source: apprunner.Source.fromEcrPublic({
+      imageConfiguration: {
+        startCommand: '/root/start-command.sh',
+      },
+      imageIdentifier: 'public.ecr.aws/aws-containers/hello-app-runner:latest',
+    }),
+  });
+
+  service.addEnvironmentVariable('TEST_ENVIRONMENT_VARIABLE', 'test environment variable value');
+  service.addSecret('LATER_SECRET', apprunner.Secret.fromSecretsManager(secret, 'field'));
```
```suggestion
  const secret = new secretsmanager.Secret(stack, 'Secret');
  const service = new apprunner.Service(stack, 'DemoService', {
    source: apprunner.Source.fromEcrPublic({
      imageConfiguration: {
        startCommand: '/root/start-command.sh',
      },
      imageIdentifier: 'public.ecr.aws/aws-containers/hello-app-runner:latest',
    }),
  });
  
  // WHEN
  service.addEnvironmentVariable('TEST_ENVIRONMENT_VARIABLE', 'test environment variable value');
  service.addSecret('LATER_SECRET', apprunner.Secret.fromSecretsManager(secret, 'field'));
```

comcalvi commented on packages/@aws-cdk/aws-apprunner/test/service.test.ts:
```diff
@@ -391,6 +391,76 @@ test('custom environment secrets and start commands are allowed for imageConfigu
 });
 
+
+test('custom environment variables and secrets can be added without first defining them in props', () => {
```
Can we split this into two tests, one for each add*() method? If someone breaks one of these methods but not another, having the name of the test match the method that was broken will make it easier to debug.

rogerchi replied:
Done

comcalvi commented on packages/@aws-cdk/aws-apprunner/lib/service.ts:
```diff
@@ -1130,36 +1132,22 @@ export class Service extends cdk.Resource {
   }
 
   private renderEnvironmentSecrets(): EnvironmentSecret[] | undefined {
-    if (Object.keys(this.environmentSecrets).length > 0 && this.instanceRole) {
-      for (const [key, value] of Object.entries(this.environmentSecrets)) {
-        if (key.startsWith('AWSAPPRUNNER')) {
-          throw new Error(`Environment secret key ${key} with a prefix of AWSAPPRUNNER is not allowed`);
-        }
-
-        value.grantRead(this.instanceRole);
-        this.secrets.push({ name: key, value: value.arn });
-      }
```
Why was this removed?

rogerchi replied:
The logic for throwing the error now happens in the .addSecret() method.  And the prop secrets are added using the same .addSecret() method now, see lines 977:985.  I followed the pattern that the `lambda.Function` construct established by using the .add* methods for both prop environment variables and ones added later using the .add* methods.

comcalvi commented on packages/@aws-cdk/aws-apprunner/lib/service.ts:
```diff
@@ -1130,36 +1132,22 @@ export class Service extends cdk.Resource {
   }
 
   private renderEnvironmentVariables(): EnvironmentVariable[] | undefined {
-    if (Object.keys(this.environmentVariables).length > 0) {
-      for (const [key, value] of Object.entries(this.environmentVariables)) {
-        if (key.startsWith('AWSAPPRUNNER')) {
-          throw new Error(`Environment variable key ${key} with a prefix of AWSAPPRUNNER is not allowed`);
-        }
-        this.variables.push({ name: key, value: value });
-      }
```
Why was this removed?

rogerchi replied:
See comment above

comcalvi commented on packages/@aws-cdk/aws-apprunner/lib/service.ts:
```diff
@@ -981,17 +972,22 @@ export class Service extends cdk.Resource {
     this.accessRole = (this.source.imageRepository?.imageRepositoryType == ImageRepositoryType.ECR) ?
       this.props.accessRole ?? this.generateDefaultRole() : undefined;
 
-    // generalte an IAM role only when environmentSecrets has values and props.instanceRole is undefined
-    this.instanceRole = (Object.keys(this.environmentSecrets).length > 0 && !this.props.instanceRole) ?
-      this.createInstanceRole() : this.props.instanceRole;
-
```
This same logic now needs to be performed lazily, no? Not entirely removed

rogerchi replied:
This is handled in lines 975, 1049, and then lazily produced in line 1000.

rogerchi commented:
@comcalvi I've tried to address/answer all of your comments.  Let me know if there's any further suggestions or clarifications required!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 2904f2d15554ae0a2eaf5f49934b04ac63b46e5e
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f55efd57-4985-409c-9831-6f1dc260e8f6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: f7a56a91b6e382789ad4b46ef759985ba9f1d627
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c3a328c9-d4aa-45b7-89a0-39dfe1803b90%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: b0e8b261272f8a87addfbbe49fb288456491f0a0
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b8ed3eb3-8cd0-4301-a08b-eff5a0e97466%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: d43fdc71b39dd88931e1cbc38a788986bb6df6ce
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bcb17c59-11f8-44e6-af17-e6ee39fcfd6e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: e9ebaf0d10c468d784860279536a59c19f132995
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=55f93646-e3ef-4724-8380-d200c232161e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

comcalvi reviewed (approved):
nice work!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: dd23bb4cfae6325426f3dfb3aa888759c17a919a
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2750590a-f1ff-4d2e-9d8e-d6494632b911%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
